Source/Data/HACompressedEntity.swift

```diff
@@ -0,0 +1,58 @@
+import Foundation
+
+public struct HACompressedStatesUpdates: HADataDecodable {
+    public var add: [String: HACompressedEntityState]?
+    public var remove: [String]?
+    public var change: [String: CompressedEntityDiff]?
+
+    public init(data: HAData) throws {
+        self.add = try? data.decode("a")
+        self.remove = try? data.decode("r")
+        self.change = try? data.decode("c")
+    }
+}
+
+public struct HACompressedEntityState: HADataDecodable {
+    public var state: String
+    public var attributes: [String: Any]?
+    public var context: String?
+    public var lastChanged: Date?
+    public var lastUpdated: Date?
+
+    public init(data: HAData) throws {
+        self.state = try data.decode("s")
+        self.attributes = try? data.decode("a")
+        self.context = try? data.decode("c")
+        self.lastChanged = try? data.decode("lc")
+        self.lastUpdated = try? data.decode("lu")
+    }
+
+    func asEntity(entityId: String) throws -> HAEntity {
+        try HAEntity(
+            entityId: entityId,
+            state: state,
+            lastChanged: lastChanged ?? Date(),
+            lastUpdated: lastUpdated ?? Date(),
+            attributes: attributes ?? [:],
+            context: .init(id: context ?? "", userId: nil, parentId: nil)
+        )
+    }
+}
+
+public struct HACompressedEntityStateRemove: HADataDecodable {
+    public var attributes: [String]?
+
+    public init(data: HAData) throws {
+        self.attributes = try? data.decode("a")
+    }
+}
+
+public struct CompressedEntityDiff: HADataDecodable {
```

Comment:
Do many of these need to be public?

Also: prefix

Reply:
Yes because `HACompressedStatesUpdates` needs to be public so `HATypedSubscription.subscribeEntities `can also be public